### PR TITLE
Emit declaration of class extending intersection

### LINKED ISSUE #14075: Allow inheritance from intersections
Based on #14017, when exporting a class that extends from a mixin, the emitted type contains an intersection. Inheriting from an intersection is not allowed, even though it should be.

Note that inheriting from a type alias of an intersection already works:

```ts
// A *self-contained* demonstration of the problem follows...
class A { a = 1 } 
class B { b = 2 }
type Both = A & B
class C extends A & B { } // error
class D extends Both { } // ok
```

**Expected behavior:**
No error for C *or* D.

**Actual behavior:**
Error for C.

### PR DESCRIPTION
**Pull request by sandersn**
Classes that extend expressions will get a synthetic var declaration for
the expression. This is required for classes that extend an expression
that return an intersection type.

Fixes #14075, but *not* #15066, which is the full mixin scenario. You still can't export a class that inherits from a class expression that was returned from a function. That probably requires more complicated type declarations to be generated.

### CODE REVIEW

File: src/compiler/declarationEmitter.ts [hunk 1]
```diff
@@ -1151,21 +1154,39 @@ namespace ts {
                 }
             }
 
+            const prevEnclosingDeclaration = enclosingDeclaration;
+            enclosingDeclaration = node;
+            const baseTypeNode = getClassExtendsHeritageClauseElement(node);
+            let tempVarName: string;
+            if (isNonNullExpression(baseTypeNode)) {
+                tempVarName = emitTempVariableDeclaration(baseTypeNode.expression, `_${node.name.text}_intersection_base`, {
```
Comment (mhegazy) [inline src/compiler/declarationEmitter.ts]: i would remove the `_intersection`

File: src/compiler/declarationEmitter.ts [hunk 2]
```diff
@@ -613,24 +612,31 @@ namespace ts {
             }
         }
 
+        function emitTempVariableDeclaration(expr: Expression, baseName: string, diagnostic: SymbolAccessibilityDiagnostic): string {
+            const tempVarName = getExportTempVariableName(baseName);
+            if (!noDeclare) {
+                write("declare ");
+            }
+            write("var ");
```
Comment (mhegazy) [inline src/compiler/declarationEmitter.ts]: consider making this `const`

File: src/compiler/declarationEmitter.ts (hunk 1 above)
Comment (mhegazy) [inline src/compiler/declarationEmitter.ts]: do not think you need the leading "_" either

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -22645,7 +22645,7 @@ namespace ts {
             const classType = <InterfaceType>getDeclaredTypeOfSymbol(getSymbolOfNode(node));
             resolveBaseTypesOfClass(classType);
             const baseType = classType.resolvedBaseTypes.length ? classType.resolvedBaseTypes[0] : unknownType;
-            if (!baseType.symbol) {
+            if (!baseType.symbol && !(baseType.flags & TypeFlags.Intersection)) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: this should not be necessary here, right? we never call this function with an expression..

**Comment by mhegazy** [comment]
@sandersn, does this also address https://github.com/Microsoft/TypeScript/issues/15572?

**Comment by mhegazy** [comment]
Please port the change to release-2.3 as well.

**Comment by sandersn** [comment]
It should fix #15572 since the error arises from incorrect flattening of an expression in `extends`. I'm testing it right now.

**Comment by sandersn** [comment]
Yes, it fixes #15572. I also ported this change to 2.3

**Pull request merged**